### PR TITLE
Move the detection of cuDNN to FindCUDNN.cmake

### PR DESCRIPTION
**Pull request by xuhdev**
Currently they sit together with other code in cuda.cmake. This commit
is the first step toward cleaning up cuDNN detection in our build system.

Another attempt to #24293,  which breaks manywheels build because it does not handle `USE_STATIC_CUDNN`.

### CODE REVIEW

**Comment by xuhdev** [comment]
The first commit is the real content, and the second is CI tweak.

**Comment by ezyang** [comment]
Good to go assuming CI passes. Will need to re-disable the CI tweaks before landing

**Comment by xuhdev** [comment]
Do we intend to support static cuDNN combined with dynamic CUDA? I'm wondering about this because, even before this change, `CAFFE2_STATIC_LINK_CUDA` and `USE_STATIC_CUDNN` coexist but `CAFFE2_STATIC_LINK_CUDA` seems to be overriding `USE_STATIC_CUDNN`:
https://github.com/pytorch/pytorch/blob/561037aef8134d81479fac6cdbd196d3652fd8e7/cmake/public/cuda.cmake#L75-L79

**Comment by ezyang** [comment]
> Do we intend to support static cuDNN combined with dynamic CUDA?

@soumith would know for sure, but that sounds pretty nutty to me. The reason we statically link cuda/cudnn is so we can distribute wheels without assuming that the user has an appropriately setup cuda. Having one statically linked but not the other sounds nutso.

**Comment by xuhdev** [comment]
If we do not need to support such a combination, we can merge all these options at issue to one single `USE_STATIC_CUDA`.

**Comment by xuhdev** [comment]
@ezyang the CI commit is reverted now.

**Review by ezyang — APPROVED** [review]

**Comment by soumith** [comment]
> Do we intend to support static cuDNN combined with dynamic CUDA?

Absolutely. In fact, we use this commonly for `conda` binaries, where `cudatoolkit` is used from anaconda, but `cudnn` is statically linked

**Comment by ezyang** [comment]
Breaks binary build on binary_linux_manywheel_3.7m_cu100_devtoolset7_test

```
Aug 20 10:34:57 + echo 'All dependencies of /opt/python/cp37-cp37m/bin/../lib/python3.7/site-packages/torch/_C.cpython-37m-x86_64-linux-gnu.so are 	linux-vdso.so.1 =>  (0x00007ffc78f98000)
Aug 20 10:34:57 	libshm.so => /opt/python/cp37-cp37m/bin/../lib/python3.7/site-packages/torch/lib/libshm.so (0x00007f5919f04000)
Aug 20 10:34:57 	libtorch_python.so => /opt/python/cp37-cp37m/bin/../lib/python3.7/site-packages/torch/lib/libtorch_python.so (0x00007f5919167000)
Aug 20 10:34:57 	libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f5918e60000)
Aug 20 10:34:57 	libm.so.6 => /usr/lib64/libm.so.6 (0x00007f5918b5e000)
Aug 20 10:34:57 	libgcc_s.so.1 => /usr/lib64/libgcc_s.so.1 (0x00007f5918948000)
Aug 20 10:34:57 	libpthread.so.0 => /usr/lib64/libpthread.so.0 (0x00007f591872c000)
Aug 20 10:34:57 	libc.so.6 => /usr/lib64/libc.so.6 (0x00007f591835f000)
Aug 20 10:34:57 	libtorch.so => /opt/python/cp37-cp37m/bin/../lib/python3.7/site-packages/torch/lib/libtorch.so (0x00007f58dace6000)
Aug 20 10:34:57 	librt.so.1 => /usr/lib64/librt.so.1 (0x00007f58daade000)
Aug 20 10:34:57 	libdl.so.2 => /usr/lib64/libdl.so.2 (0x00007f58da8da000)
Aug 20 10:34:57 	libc10_cuda.so => /opt/python/cp37-cp37m/bin/../lib/python3.7/site-packages/torch/lib/libc10_cuda.so (0x00007f58da6b0000)
Aug 20 10:34:57 	libc10.so => /opt/python/cp37-cp37m/bin/../lib/python3.7/site-packages/torch/lib/libc10.so (0x00007f58da459000)
Aug 20 10:34:57 	libnvToolsExt-3965bdd0.so.1 => /opt/python/cp37-cp37m/bin/../lib/python3.7/site-packages/torch/lib/libnvToolsExt-3965bdd0.so.1 (0x00007f58da24f000)
Aug 20 10:34:57 	libcudart-1581fefa.so.10.0 => /opt/python/cp37-cp37m/bin/../lib/python3.7/site-packages/torch/lib/libcudart-1581fefa.so.10.0 (0x00007f58d9fd2000)
Aug 20 10:34:57 	libcudnn.so.7 => not found
Aug 20 10:34:57 	libgomp-4f651535.so.1 => /opt/python/cp37-cp37m/bin/../lib/python3.7/site-packages/torch/lib/libgomp-4f651535.so.1 (0x00007f58d9da8000)
Aug 20 10:34:57 	/lib64/ld-linux-x86-64.so.2 (0x00007f591a32e000)
Aug 20 10:34:57 	libcudnn.so.7 => not found with runpath   RUNPATH              $ORIGIN:$ORIGIN/lib'
Aug 20 10:34:57 + [[ manywheel == \l\i\b\t\o\r\c\h ]]
Aug 20 10:34:57 + python -c 'import torch'
Aug 20 10:34:57 Traceback (most recent call last):
Aug 20 10:34:57   File "<string>", line 1, in <module>
Aug 20 10:34:57   File "/opt/python/cp37-cp37m/lib/python3.7/site-packages/torch/__init__.py", line 81, in <module>
Aug 20 10:34:57     from torch._C import *
Aug 20 10:34:57 ImportError: libcudnn.so.7: cannot open shared object file: No such file or directory
```

https://circleci.com/gh/pytorch/pytorch/2514905?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

**Comment by ezyang** [comment]
I think the reason we missed this in CI, is because you also have to edit the `should_run.py` script. You can see the job aborted early because it decided it didn't need to run: https://circleci.com/gh/pytorch/pytorch/2490176?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by xuhdev** [comment]
You are right. Let me rework on this

**Pull request closed without merging**
